scylla-cql/src/_macro_internal.rs

```diff
@@ -64,5 +104,84 @@ pub mod ser {
+        ///   f1: i32, // ...serialization of `f1` will return `FieldStatus::NotDone`...
+        ///   f2: f64, // ...and serialization of `f2` - `FieldStatus::Done`.
+        /// }
+        ///
+        /// #[derive(SerializeRow)]
+        /// # #[scylla(crate = scylla_cql)]
+        /// struct B {
+        ///   #[scylla(flatten)]
+        ///   a: A,
+        ///   i: i32,
+        /// }
+        /// ```
+        ///
+        /// For now this enum is an implementation detail of `#[derive(SerializeRow)]` when
+        /// serializing by name
+        #[derive(Debug)]
+        pub enum FieldStatus {
+            /// Serializing to the column finished serialization for this (possibly flattened)
+            /// field.
+            Done,
+            /// Serialization to the column was done, but there are other subfields (fields of the
+            /// flattened struct) that need to be serialized.
+            NotDone,
+            /// The column did not correspond to this field (nor its subfields, if it's flattened).
+            NotUsed,
+        }
+
+        /// Wrapper around a struct that can be serialized by name for a whole row
+        ///
+        /// Implementation detail of `#[derive(SerializeRow)]` when serializing by name
+        pub struct ByName<'t, T: SerializeRowByName>(pub &'t T);
+
+        impl<T: SerializeRowByName> ByName<'_, T> {
+            #[inline]
+            /// Serializes all the fields/columns by name
+            pub fn serialize(
+                self,
+                ctx: &RowSerializationContext,
+                writer: &mut RowWriter<'_>,
+            ) -> Result<(), SerializationError> {
+                // 1. create the partial view of the row we are serializing. A partial contains
+                // references to each serializable field and tracks which fields have already been
+                // serialized and which ones are missing
+                let mut partial = self.0.partial();
+
+                for spec in ctx.columns() {
+                    // 2. For each column attempt to serialize it using the partial view
+                    let serialized = partial.serialize_field(spec, writer)?;
+
+                    // 3. If the field was not used that means the column doesn't belong to this
+                    // struct and thus cannot be serialized. Return error.
+                    if matches!(serialized, FieldStatus::NotUsed) {
+                        return Err(mk_typck_err::<Self>(
+                            BuiltinTypeCheckErrorKind::NoColumnWithName {
+                                name: spec.name().to_owned(),
+                            },
+                        ));
+                    }
+                }
+
+                // 4. After all the fields are serialized, check that the partial doesn't have any
+                // fields left to serialize - return an error otherwise as we are missing columns
+                partial.check_missing()
+            }
+        }
```

Comment:
Ok here is where my question above starts to matter. Are there other reasons besides empty structs why we call `check_missing` instead of remembering of `Done` was returned after the last column?

Reply:
Sorry, I phrased it weird. I should have said:

The first call to check_missing **could be made** only if we didn't detect a Done, but that call to check_missing may call to flattened struct's check_missing if the field that it detected was incomplete was a flattened struct.

So yes, we can have an `if` check to only call it if we didn't detect a `Done`. But as I said in another comment, the happy case for that function is very cheap (just an `if num_remaining == 0`) so we'd be trading an if for another 🤷🏾